packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts

```diff
@@ -0,0 +1,126 @@
+import type { QueryBuilderState } from './QueryBuilderState.js';
+
+export class QueryBuilderChangeHistoryState {
+  readonly queryBuilderState: QueryBuilderState;
+  readonly initState = ActionState.create();
+
+  querySnapshotBuffer: RawLambda[] = [];
+  currentQuery: RawLambda | undefined;
+  pointer = -1;
+  bufferSize = 10;
+
+  constructor(queryBuilderState: QueryBuilderState) {
+    makeObservable(this, {
+      currentQuery: observable,
+      pointer: observable,
+      querySnapshotBuffer: observable,
+      initialize: action,
+      undo: action,
+      redo: action,
+      setCurrentQuery: action,
+      cacheNewQuery: action,
+      canRedo: computed,
+      canUndo: computed,
+    });
+
+    this.queryBuilderState = queryBuilderState;
+  }
+
+  get canRedo(): boolean {
+    return this.pointer < this.querySnapshotBuffer.length - 1;
+  }
+
+  get canUndo(): boolean {
+    return this.pointer > 0;
+  }
+
+  redo(): void {
+    if (this.canRedo) {
+      this.pointer = this.pointer + 1;
+      const query = this.querySnapshotBuffer[this.pointer];
+      this.setCurrentQuery(query);
+      this.queryBuilderState.rebuildWithQuery(guaranteeNonNullable(query), {
+        preserveParameterValues: true,
+        preserveResult: true,
+      });
+    }
+  }
+
+  undo(): void {
+    if (this.canUndo) {
+      this.pointer = this.pointer - 1;
+      const query = this.querySnapshotBuffer[this.pointer];
+      this.setCurrentQuery(query);
+      this.queryBuilderState.rebuildWithQuery(guaranteeNonNullable(query), {
+        preserveParameterValues: true,
+        preserveResult: true,
+      });
+    }
+  }
+
+  setCurrentQuery(query: RawLambda | undefined): void {
+    this.currentQuery = query;
+  }
+
+  initialize(initialQuery: RawLambda): void {
+    if (this.queryBuilderState.isQuerySupported) {
+      this.initState.inProgress();
+      this.currentQuery = initialQuery;
+      this.querySnapshotBuffer.push(initialQuery);
+      this.pointer = this.pointer + 1;
+      this.initState.complete();
+    }
+  }
+
+  cacheNewQuery(query: RawLambda): void {
+    try {
+      if (this.queryBuilderState.isQuerySupported) {
+        if (query.hashCode !== this.currentQuery?.hashCode) {
+          if (
+            this.querySnapshotBuffer.length === this.bufferSize &&
+            this.pointer === this.querySnapshotBuffer.length - 1
+          ) {
+            // only record 10 query snapshots
+            this.querySnapshotBuffer = this.querySnapshotBuffer.slice(1);
+          } else {
+            this.querySnapshotBuffer = this.querySnapshotBuffer.slice(
+              0,
+              this.pointer + 1,
+            );
+          }
+          this.querySnapshotBuffer.push(query);
+          this.pointer = this.querySnapshotBuffer.length - 1;
+          this.setCurrentQuery(query);
+        }
+      }
+    } catch (error) {
+      assertErrorThrown(error);
+      this.queryBuilderState.applicationStore.notificationService.notifyError(
```

Comment:
we should log the error not notify i think. create a new legend query. CHANGE_HISTORY_ERROR or something like that